src/lib/optimizer/strategy/subquery_to_join_rule.cpp

```diff
@@ -0,0 +1,573 @@
+
+      if (sub_expression->type == ExpressionType::LQPSubquery) {
+        // Need to check whether the subquery uses correlated parameters
+        const auto& lqp_select_expression = std::static_pointer_cast<LQPSubqueryExpression>(sub_expression);
+        visit_lqp(lqp_select_expression->lqp, [&](const auto& sub_node) {
+          if (is_correlated) {
+            return LQPVisitation::DoNotVisitInputs;
+          }
+
+          is_correlated |= uses_correlated_parameters(sub_node, parameter_mapping);
+          return is_correlated ? LQPVisitation::DoNotVisitInputs : LQPVisitation::VisitInputs;
+        });
+      } else if (sub_expression->type == ExpressionType::CorrelatedParameter) {
+        const auto& parameter_expression = std::static_pointer_cast<CorrelatedParameterExpression>(sub_expression);
+        if (parameter_mapping.find(parameter_expression->parameter_id) != parameter_mapping.end()) {
+          is_correlated = true;
+        }
+      }
+
+      return is_correlated ? ExpressionVisitation::DoNotVisitArguments : ExpressionVisitation::VisitArguments;
+    });
+
+    if (is_correlated) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Searches for usages of correlated parameters that we cannot optimize.
+ *
+ * This includes two things:
+ *   - Usages of correlated parameters outside of predicate nodes (for example joins)
+ *   - Usages of correlated parameters in predicate nodes nested below nodes they cannot be pulled up past.
+ */
+bool contains_unoptimizable_correlated_parameter_usages(
+    const std::shared_ptr<AbstractLQPNode>& lqp,
+    const std::map<ParameterID, std::shared_ptr<AbstractExpression>>& parameter_mapping,
+    const std::set<std::shared_ptr<PredicateNode>> optimizable_predicate_nodes) {
+  bool optimizable = true;
+  visit_lqp(lqp, [&](const auto& node) {
+    if (!optimizable) {
+      return LQPVisitation::DoNotVisitInputs;
+    }
+
+    auto correlated = uses_correlated_parameters(node, parameter_mapping);
+    if (correlated) {
+      if (node->type != LQPNodeType::Predicate ||
+          optimizable_predicate_nodes.find(std::static_pointer_cast<PredicateNode>(node)) ==
+              optimizable_predicate_nodes.end()) {
+        optimizable = false;
+        return LQPVisitation::DoNotVisitInputs;
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  return !optimizable;
+}
+
+/**
+ * Replace aggregate nodes by ones with additional group by expressions added to allow predicates to be pulled up.
+ */
+void replace_aggregate_nodes(PredicatePullUpInfo& pull_up_info) {
+  // Walk the predicates from bottom to top while keeping track of the column expressions they need to be available.
+  // Whenever an aggregate node is encountered, replace it by a new one with the column expressions of all the
+  // predicates below it added as group by expressions.
+
+  // TODO(anybody): This code does quite a bit of unnecessary copying around because the AggregateNode constructor
```

Comment:
The AggregateNode is not going to change in that regard, pls adjust the rule and/or remove this TODO